cabot/cabotapp/models.py

```diff
@@ -1146,6 +1169,146 @@ def __unicode__(self):
+
+        if self.match_if == self.MATCH_CHECK:
+            return True
+
+        raise NotImplementedError()
+
+
+class Acknowledgement(ResultFilter):
+    """
+    A filter that matches on a StatusCheck (or subset of a StatusCheck, if matching on tags).
+    An Acknowledgement is considered 'open' if closed_at is set to null.
+
+    Acknowledgements "expire" (auto-close) if the expire_at parameter is set. Expiration is performed in two ways:
+      1) get_acks_matching_check() will not return checks where expire_at <= now.
+      2) A periodic celery task calls ack.close() on checks where expire_at <= now.
+
+    Acknowledgements are also automatically closed when their StatusCheck succeeds at least close_after_successes
+    times (consecutively). This is done by Acknowledgement.close_succeeding_acks(), which is called by StatusCheck.run()
+    whenever a check succeeds.
+
+    For simplicity, only one Acknowledgement can exist per StatusCheck. This is enforced by automatically closing
+    Acknowledgements that already exist for the same status check in Acknowledgement.save() (with a fixed reason).
+    """
+    created_at = models.DateTimeField(default=timezone.now, db_index=True)
+    created_by = models.ForeignKey(User, null=True, default=None)
+    note = models.TextField(max_length=255, blank=True,
+                            help_text='Leave a note explaining why this ack was created.')
+
+    # if 'closed_at' is set, the ack is considered 'closed'
+    closed_at = models.DateTimeField(null=True, default=None, db_index=True)
+    closed_reason = models.TextField(max_length=255, null=True, blank=False, default=None)
+
+    expire_at = models.DateTimeField(null=True, default=None, db_index=True,
+                                     help_text='After this time the acknowledgement will be automatically closed and '
+                                               'alerts will resume, even if the check is still failing.')
+    close_after_successes = models.PositiveIntegerField(null=True, default=1,
+                                                        help_text='After this many consecutive successful runs the '
+                                                                  'acknowledgement will be automatically closed. Enter '
+                                                                  '0 to disable.')
+
+    @classmethod
+    def get_acks_matching_check(cls, check, at_time=None):
+        # type: (StatusCheck, timezone.datetime) -> List[Acknowledgement]
+        """
+        :param check: check to gather acks for
+        :param at_time: only consider acks that were open at this time; leave as None to use timezone.now().
+        :returns list of Acknowledgements where ack.
+        """
+        at_time = at_time or timezone.now()
+        acks = cls.objects.filter(status_check_id=check.id)
+        return acks.exclude(created_at__gt=at_time).exclude(closed_at__lte=at_time).exclude(expire_at__lte=at_time)
```

Comment:
This could, at worst, be dealing with 60 days worth of acks - which I expect would be in the hundreds at worst. We should probably add some indices...

Reply:
yeah I think an index would be good form